doc/developer/chips/10.rst

```diff
@@ -531,24 +531,76 @@ For more details on when the copy initializer would be called, please refer to
 Deinitialization
 ++++++++++++++++
 
-Records and classes can specify actions to take when a variable of that type is
-deinitialized. Record variables are deinitialized when they go out of scope and
-class variables are deinitialized when `delete` is called on them.
+Deinitialization is performed in these cases:
 
-To specify additional actions that occur upon deinitialization, specify a `proc
+ * Record variables are deinitialized when they go out of scope.
+ * Class variables are deinitialized when `delete` is called on them.
+
+To specify actions that occur upon deinitialization, user can specify a `proc
 deinit()` method. This method must have zero arguments and cannot be called
 directly. If a type does not specify a `proc deinit()` method, the compiler
 will generate one.
+It will be equivalent to a user-written `deinit` with an empty body.
+
+In addition to the actions specified by the `deinit` method, whether
+compiler-generated or user-defined, in the process of deinitialization:
 
-No matter whether the `deinit` method is compiler generated or user-defined,
-the compiler will ensure that in the process of deinitialization, all fields of
-record type are deinitialized. Additionally, the compiler will ensure that
-parent `deinit` methods are called in the case of inherited classes.
+ * parent `deinit` methods are called in the case of inherited classes, and
+ * all fields of record type are deinitialized.
 
 The expected use-case for the `deinit` method is to call `delete` on
 fields of class type, when appropriate. It is the responsibility of the
 class designer to `delete` fields when appropriate. The compiler will not
 automatically add `delete` calls for fields of class type.
+Conversely, the compiler adds `deinit` calls on fields of record type.
+These calls cannot be added by the user.
+
+Ensuring type safety
+********************
```

Comment:
I think this section is a good discussion of our rationale, but I never intended to include such a detailed rationale in this CHIP. If we decide to include the Rationale here, we should put it after the description of the language rules. I think it's confusing here because a user might interpret "could potentially be accessed" as saying that it is legal Chapel to write code matching one of these sections.

In particular, it was our intention to prevent a user from using the whole object `this` because dynamic dispatch could occur. We relaxed that slightly to allow field access (because these are technically methods now) but that was more an accommodation to the current compiler implementation than anything else. I'm not sure we agreed that you could pass `this` to a function that never does dynamic dispatch on it.

Lastly, in the case of section (d), we can make "dependencies" between fields in any order we please with mutable fields. I havn't personally been very satisfied with the term "dependency" in this context and I'm not sure what it means. Consider this:

``` chapel
class MyObject { }
record Owned {
 var instance;
 proc deinit() { delete instance; }
}
class Parent {
  var myInstance: MyObject;
  proc init() {
    myInstance = new MyObject(); 
  }
  proc deinit() {
    writeln(myInstance);
  }
}
class Child {
  var pointsToParent: Owned(MyObject);
  proc init() {
    pointsToParent = new Owned(nil);
    super.init();
    pointsToParent.instance = this.myInstance;
    // now pointsToParent refers to the object in Parent.myInstance
  }
}
var x = new Child();
delete x;
```

Would you call pointsToParent a "dependency" ?
Uh-oh! The deinitialization process on `x` will cause Parent.myInstance to be deleted before it is used in Parent.deinit!

Or consider this example, which has "dependent" fields in reverse declaration order:
``` chapel
record ObjectPair {
  var a: MyObject;
  var b: MyObject;
  proc deinit() {
    writeln(a);
    writeln(b);
  }
}
class Test {
  var recPair: ObjectPair;
  var objA:Owned(MyObject);
  var objB:Owned(MyObject);
  proc init() {
    super.init();
    objA = new Owned(new MyObject());
    objB = new Owned(new MyObject());
    recPair = new ObjectPair( objA, objB );
  }
}
var y = new Test();
delete y;
```

Would you call recPair's use of objA and objB a "dependency"?
Uh-oh! The deinitialization process on `y` will deinitialize objA and objB  before they are used in `ObjectPair.deinit`.

Bottom line: I don't think we can ever really "ensure type safety" as long as `delete` is subject to user management (and not checked to a great degree by the compiler). I think the rules we had in the CHIP already are fine (along with your minor changes to them) but I don't agree with this new section.